lib/vnet/ssh_proxy.go

```diff
@@ -0,0 +1,254 @@
+	runTask(func() {
+		proxyChannels(ctx, serverConn.conn, clientConn.chans, closeConnections)
+	})
+	runTask(func() {
+		proxyChannels(ctx, clientConn.conn, serverConn.chans, closeConnections)
+	})
+
+	// Proxy global requests in both directions.
+	runTask(func() {
+		proxyGlobalRequests(ctx, serverConn.conn, clientConn.reqs, closeConnections)
+	})
+	runTask(func() {
+		proxyGlobalRequests(ctx, clientConn.conn, serverConn.reqs, closeConnections)
+	})
+
+	wg.Wait()
+}
+
+func proxyChannels(
+	ctx context.Context,
+	targetConn ssh.Conn,
+	chans <-chan ssh.NewChannel,
+	closeConnections func(),
+) {
+	// Proxy each SSH channel in its own goroutine, make sure they don't leak by
+	// tracking with a WaitGroup.
+	var wg sync.WaitGroup
+	for newChan := range chans {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			proxyChannel(ctx, targetConn, newChan, closeConnections)
+		}()
+	}
+	wg.Wait()
+}
+
+func proxyChannel(
+	ctx context.Context,
+	targetConn ssh.Conn,
+	newChan ssh.NewChannel,
+	closeConnections func(),
+) {
+	log := log.With("channel_type", newChan.ChannelType())
+	log.DebugContext(ctx, "Proxying new SSH channel")
+
+	// Try to open a corresponding channel on the target.
+	targetChan, targetChanRequests, err := targetConn.OpenChannel(
+		newChan.ChannelType(), newChan.ExtraData())
+	if err != nil {
+		var openChannelErr *ssh.OpenChannelError
+		var rejectErr error
+		if errors.As(err, &openChannelErr) {
+			// The target rejected the channel, this is totally expected in some
+			// cases, just reject the incoming channel request.
+			rejectErr = trace.Wrap(newChan.Reject(openChannelErr.Reason, openChannelErr.Message))
+		} else {
+			// We got an unexpected error trying to open the channel on the
+			// target, this is fatal, log and kill the connection.
+			log.DebugContext(ctx, "Unexpected error opening SSH channel on target",
+				"error", err)
+			msg := "unexpected error opening channel on target: " + err.Error()
+			rejectErr = trace.Wrap(newChan.Reject(ssh.ConnectionFailed, msg))
+			closeConnections()
+		}
+		if rejectErr != nil {
+			// We failed to reject the incoming channel, this is fatal, log and
+			// kill the connection.
+			log.DebugContext(ctx, "Failed to reject SSH channel request",
+				"error", err)
+			closeConnections()
+		}
+		return
+	}
+
+	// Now that the target accepted the channel, accept the incoming channel
+	// request.
+	incomingChan, incomingChanRequests, err := newChan.Accept()
+	if err != nil {
+		// Failing to accept an incoming channel request that the target already
+		// accepted is fatal. Log, close the channel we just opened on the
+		// target, and kill the connection.
+		log.DebugContext(ctx, "Failed to accept SSH channel request already accepted by the target, killing the connection",
+			"error", err)
+		if err := targetChan.Close(); err != nil {
+			log.DebugContext(ctx, "Failed to close SSH channel on target",
+				"error", err)
+		}
+		closeConnections()
```

Comment:
Same as above, a fatal error on one side of the connection should fatally kill the other side.

```suggestion
		closeConnections()
		go ssh.DiscardRequests(targetChanRequests)
		_ = targetChan.Close()
```

Reply:
i did forget to drain the channel here, good catch, fixed